[docs/data] Add Blocks and Streaming Execution to Quickstart (#50022)
ray-project/ray: doc/source/data/quickstart.rst

Quickstart
==========

> This page introduces the :class:`Dataset <ray.data.Dataset>` concept and the capabilities it provides.

Reviewer: nit: but perhaps best to state this is a guide to get up and running with Ray Data instead of "about the Dataset concept"

* :ref:`Streaming Execution Model <streaming_execution_model>`

.. _data_key_concepts:

Key Concepts: Datasets and Blocks
---------------------------------

There are two main concepts in Ray Data:

* :class:`Dataset <ray.data.Dataset>`
* :ref:`Blocks <blocks_key_concept>`

:class:`Dataset <ray.data.Dataset>` is the main user-facing Python API. It represents a
distributed data collection and defines data loading and processing operations. You
typically use the API in this way:

1. Create a Dataset from external storage or in-memory data.
2. Apply transformations to the data.
3. Write the outputs to external storage or feed the outputs to training workers.

The Dataset API is lazy, meaning that operations are not executed until you call an action
like :meth:`~ray.data.Dataset.show`. This allows Ray Data to optimize the execution plan
and execute operations in parallel.

A *block* is the basic unit of data that Ray Data operates on. A block is a contiguous
subset of rows from a dataset.

The following figure visualizes a dataset with three blocks, each holding 1000 rows.
Ray Data holds the :class:`~ray.data.Dataset` on the process that triggers execution

> (which is usually the driver) and stores the blocks as objects in Ray's shared-memory

Reviewer 2: maybe link to driver definition in glossary? New ray user may not know exactly what it is

To learn more about saving dataset contents, see :ref:`Saving data <saving-data>`.

.. _streaming_execution_model:

> Streaming Execution Model

Reviewer 3: ```suggestion
Streaming execution model
```

-------------------------

Ray Data uses a streaming execution model to efficiently process large datasets.
Rather than materializing the entire dataset in memory at once,
Ray Data can process data in a streaming fashion through a pipeline of operations.

Here's how it works:

.. code-block:: python

    import ray

    # Create a dataset with 1K rows
    ds = ray.data.range(1_000)

    # Define a pipeline of operations
    ds = ds.map(lambda x: x * 2)
    ds = ds.filter(lambda x: x % 4 == 0)

    # Data starts flowing when you call an action like show()
    ds.show(5)

Key benefits of streaming execution include:

> * **Memory Efficient**: Processes data in chunks rather than loading everything into memory

Reviewer: nitpick but best stick to blocks and not introduce a new term

* **Pipeline Parallelism**: Different stages of the pipeline can execute concurrently
* **Automatic Memory Management**: Ray Data automatically spills data to disk if memory pressure is high

> * **Lazy Evaluation**: Transformations are not executed until an action triggers the pipeline

Reviewer 3: https://developers.google.com/style/contractions
```suggestion
* **Lazy Evaluation**: Transformations aren't executed until an action triggers the pipeline
```

The streaming model allows Ray Data to efficiently handle datasets much larger than memory while maintaining high performance through parallel execution.

.. note::

>    Operations like :meth:`ds.sort() <ray.data.Dataset.sort>` and :meth:`ds.groupby() <ray.data.Dataset.groupby>` require materializing data, which may impact memory usage for very large datasets.

Reviewer: nitpick: best to refer to methods that are decorated as consumption API (e.g. `ds.split` or `ds.split_at_indices`) as opposed to all to all operations like `groupby` which theoretically don't require a full materializing of the data
